### PR TITLE
Icseg cases

### PR DESCRIPTION
**Pull request by jbarberia**
grg-pssedata fails at reading the case collection of https://icseg.iti.illinois.edu/power-cases/

- I put some defaults arguments in the `struct.py`  classes.
- I changed how the `.to_psse()` method write the case to avoid the excesive blank lines.

### CODE REVIEW

File: grg_pssedata/struct.py [hunk 1]
```diff
@@ -983,7 +994,7 @@ def validate(self):
     def to_psse(self):
         '''Returns: a pss/e encoding of this data structure as a string'''
         data = [self.p1, self.p2, self.w1, self.w2]
-        return os.linesep.join([x.to_psse() for x in data])
+        return '\n'.join([x.to_psse() for x in data])
```
Comment (ccoffrin) [inline grg_pssedata/struct.py:998]: This change is interesting, I would expect it to be less robust. Can you elaborate on this one?

File: grg_pssedata/struct.py [hunk 2]
```diff
@@ -1037,13 +1048,13 @@ def validate(self):
     def to_psse(self):
         '''Returns: a pss/e encoding of this data structure as a string'''
         data = [self.p1, self.p2, self.w1, self.w2, self.w3]
-        return os.linesep.join([x.to_psse() for x in data])
+        return '\n'.join([x.to_psse() for x in data])
```
Comment (ccoffrin) [inline grg_pssedata/struct.py:1052]: Same as above.

File: grg_pssedata/struct.py [hunk 3]
```diff
@@ -1667,7 +1678,7 @@ def validate(self):
     def to_psse(self):
         '''Returns: a pss/e encoding of this data structure as a string'''
         data = [self.params, self.c1, self.c2]
-        return os.linesep.join([x.to_psse() for x in data])
+        return '\n'.join([x.to_psse() for x in data])
```
Comment (ccoffrin) [inline grg_pssedata/struct.py:1682]: Same

File: grg_pssedata/struct.py [hunk 4]
```diff
@@ -1841,7 +1852,7 @@ def validate(self):
     def to_psse(self):
         '''Returns: a pss/e encoding of this data structure as a string'''
         data = [self.params, self.rectifier, self.inverter]
-        return os.linesep.join([x.to_psse() for x in data])
+        return '\n'.join([x.to_psse() for x in data])
```
Comment (ccoffrin) [inline grg_pssedata/struct.py:1856]: same

File: grg_pssedata/struct.py [hunk 5]
```diff
@@ -2431,7 +2442,7 @@ def to_psse(self):
         data.extend(self.converters)
         data.extend(self.dc_buses)
         data.extend(self.dc_links)
-        return os.linesep.join([x.to_psse() for x in data])
+        return '\n'.join([x.to_psse() for x in data])
```
Comment (ccoffrin) [inline grg_pssedata/struct.py:2446]: same

File: grg_pssedata/struct.py [hunk 6]
```diff
@@ -59,6 +59,17 @@ def quote_string(s):
...
 
+def _psse_str(s):
+    '''format numbers to PSSE style'''
+    if isinstance(s, float):
+        if abs(s) <= 1e-1:
+            return "{:.5e}".format(s)
+        elif 1e-1 < abs(s) < 1e1:
+            return "{:.5f}".format(s)
+        else:
+            return "{:.3f}".format(s)
```
Comment (ccoffrin) [inline grg_pssedata/struct.py]: I am generally not a fan of eliminating some floating point precision without a strong context to know it won't break the model.  Could you elaborate on what you where thinking here?

Switching between decimal and scientific does seem reasonable to me.

**Review by ccoffrin — COMMENTED** [review]
Hi @jbarberia, thanks for your contribution!  I have a few questions about some of the changes but the new proposed defaults are a welcome addition.

File: grg_pssedata/struct.py (hunk 1 above)
Comment (jbarberia) [reply grg_pssedata/struct.py:998]: There are some troubles to write the file in text mode. In Windows u get a lot of blanklines because ```os.linesep``` introduces ```\r\n``` at the end of the line. But it is interpreted as a double line break. See the [Docs Recomendations](https://docs.python.org/3/library/os.html#os.linesep).

An small example:
``` 
case = grg_pssedata.io.parse_psse_case_file('case3.raw')
f = open('foo.raw', 'w')
f.write(case.to_psse())
f.close()
``` 
With the ```master``` branch:
```
0, 100.0, 33, 0.0, 0.0, 60.0

tests basic network fetures

this case does not have an HVDC line that occurs in the Matpower case

  1, '1           ', 240.0, 3, 1, 1, 1, 1.10000002, 0.0, 1.1, 0.9, 1.1, 0.9

  2, '2           ', 240.0, 2, 1, 1, 1, 0.92616999, 7.25883, 1.1, 0.9, 1.1, 0.9

  3, '3           ', 240.0, 2, 1, 2, 1, 0.89999998, -17.267101, 1.1, 0.9, 1.1, 0.9

0 / END OF BUS DATA, BEGIN LOAD DATA
```

With the ```ICSEG-cases``` branch:

```
0, 100.0, 33, 0, 0, 60.0
tests basic network fetures
this case does not have an HVDC line that occurs in the Matpower case
  1, '1           ', 240.000, 3, 1, 1, 1, 1.10000, 0.00000e+00, 1.10000, 0.90000, 1.10000, 0.90000
  2, '2           ', 240.000, 2, 1, 1, 1, 0.92617, 7.25883, 1.10000, 0.90000, 1.10000, 0.90000
  3, '3           ', 240.000, 2, 1, 2, 1, 0.90000, -17.267, 1.10000, 0.90000, 1.10000, 0.90000
0 / END OF BUS DATA, BEGIN LOAD DATA
```

File: grg_pssedata/struct.py (hunk 6 above)
Comment (jbarberia) [reply grg_pssedata/struct.py]: I was thinking of formatting it so that it can be readable from the .RAW file in a more similar way to the .RAW file generate with the PSSE.

I agree that it may not be the best idea. We could discard this change.

File: grg_pssedata/struct.py (hunk 2 above)
Comment (jbarberia) [reply grg_pssedata/struct.py:1052]: [Reply Above](https://github.com/lanl-ansi/grg-pssedata/pull/18/commits/4837e9ab096b5a7b2750772032c1d2821a71f811#r540152037)

File: grg_pssedata/struct.py (hunk 1 above)
Comment (ccoffrin) [reply grg_pssedata/struct.py:998]: Got it, I had a miss understanding of how python uses `os.linesep` vs `'\n'`, these changes look good to me.

File: grg_pssedata/struct.py (hunk 6 above)
Comment (ccoffrin) [reply grg_pssedata/struct.py]: In the interest of getting your fixes merged soon, I propose we change this function to replicate the the previous behavior.  You can just comment out the formatting suggestion and make the function `return str(s)`.  Sound good?

Comment (jbarberia) [reply grg_pssedata/struct.py]: Sure, Sounds good to me.

**Review by jbarberia — COMMENTED** [review]
I just commented it. Or it is better to remove it?

**Comment by ccoffrin** [comment]
This is looking good to me.  I would like to get CI working before merging to make sure there are no regressions with respect the current tests.  Give me a couple days to sort that out.

**Comment by ccoffrin** [comment]
According to the [test log ](https://travis-ci.org/github/lanl-ansi/grg-pssedata/jobs/749184673) one cases is now parseable that was not previously parsed.

This one previously threw an error but now its fine, https://github.com/lanl-ansi/grg-pssedata/blob/master/tests/data/exception/powermodels/parser_test_i.raw

Does that seem reasonable to you?  If so, we need to move that case from the "test/exception" directory into "test/warning" or "test/correct".

**Comment by jbarberia** [comment]
Yes, it will be better to move it to the folder ```tests/data/correct/powermodels``` together with the other cases like ```parser_test_a.raw``` or ```parser_test_b.raw```.

Let me know what other change would be interesting to add.

File: grg_pssedata/io.py [hunk 7]
```diff
@@ -118,7 +118,7 @@ def parse_line(line, line_reqs=None):
 
     if line_reqs is not None:
         if len(line_parts) < line_reqs.min_values:
-            raise PSSEDataParsingError('on psse data line {} in the "{}" section, at least {} values were expected but only {} where found.\nparsed: {}'.format(line_reqs.line_index, line_reqs.section, line_reqs.min_values, len(line_parts), line_parts))
+            raise PSSEDataWarning('on psse data line {} in the "{}" section, at least {} values were expected but only {} where found.\nparsed: {}'.format(line_reqs.line_index, line_reqs.section, line_reqs.min_values, len(line_parts), line_parts))
```
Comment (ccoffrin) [inline grg_pssedata/io.py]: Tests are passing now, so I think we can merge this.  One last question.  Why was this changed from `PSSEDataParsingError` to `PSSEDataWarning`?  I think this change will results in a downstream error when the component data structure is constructed.

The changes you made in the `LineRequirements` should be sufficient to support the increased number of default  values.

Comment (jbarberia) [reply grg_pssedata/io.py]: I had thought at first that this solves the problem of incomplete lines. But I forgot to return it to its original state.

**Comment by ccoffrin** [comment]
Thanks for your contribution @jbarberia!

Would you like me to release a new version of the package with these changes or do you plan to propose further updates?

**Comment by jbarberia** [comment]
Thank you for all your comments!

At the moment I have no plans to add any more changes. So you can release a new version.
Maybe in the future it would be nice to have a grg-pssedata that works with PSSE 34 files.

**Comment by ccoffrin** [comment]
Ok will do.  Support for PSSE 34 would be great.  For now I am not using python that much so this is not high on my priority list, but such a contribution would be welcome.

**Comment by ccoffrin** [comment]
Release has been made.
